File: flutter/runa/lib/models/phrase.dart
import 'package:realm/realm.dart';

part 'phrase.realm.dart';

@RealmModel()
class _Phrase {
  @PrimaryKey()
  late int id;
  
  late String text;
  
  DateTime? createdAt;
  
  late String source; // 'base', 'gemini', 'firestore'
  
  bool isShown = false; // Para controlar qué frases ya fueron mostradas
  
  DateTime? lastShownAt;
}

// Clase helper para JSON
class PhraseJson {
  final int id;
  final String text;
  final DateTime? createdAt;
  final String source;
  
  PhraseJson({
    required this.id,
    required this.text,
    this.createdAt,
    required this.source,
  });
  
  factory PhraseJson.fromJson(Map<String, dynamic> json) {
    return PhraseJson(
      id: json['id'] as int,
      text: json['text'] as String,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      source: json['source'] as String,
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'text': text,
      'createdAt': createdAt?.toIso8601String(),
      'source': source,
    };
  }
  
  Phrase toRealmObject() {
    return Phrase(
      id,
      text,
      source,
      createdAt: createdAt,
    );
  }
}

File: apple/runa/runa/Models/Phrase.swift
//
//  Phrase.swift
//  runa
//
//  Created by JOSE ZARABANDA on 7/30/25.
//

import Foundation
import RealmSwift

class Phrase: Object, Identifiable {
    @Persisted var id: Int = 0
    @Persisted var text: String = ""
    @Persisted var createdAt: Date?
    @Persisted var source: String = "base" // 'base', 'gemini', 'firestore'
    @Persisted var isShown: Bool = false // Para controlar qué frases ya fueron mostradas
    @Persisted var lastShownAt: Date?
    
    override static func primaryKey() -> String? {
        return "id"
    }
    
    convenience init(id: Int, text: String, createdAt: Date?, source: String) {
        self.init()
        self.id = id
        self.text = text
        self.createdAt = createdAt
        self.source = source
    }
}

// Thread-safe data structure for passing phrase data between services
struct PhraseData {
    let id: Int
    let text: String
    let createdAt: Date?
    let source: String
    let isShown: Bool
    let lastShownAt: Date?
}

// Estructura para decodificar JSON
struct PhraseJSON: Codable, Identifiable {
    let id: Int
    let text: String
    let createdAt: String?
    let source: String
    
    var phraseObject: Phrase {
        let phrase = Phrase()
        phrase.id = id
        phrase.text = text
        phrase.source = source
        if let createdAtString = createdAt {
            phrase.createdAt = ISO8601DateFormatter().date(from: createdAtString)
        }
        return phrase
    }
}

struct PhrasesContainer: Codable {
    let phrases: [PhraseJSON]
}
